rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,653 @@
+
+We should be cautious that a BotInstance cannot be impersonated using a
+second pre-image attack. This risk is introduced by using the public key
+fingerprint as an identifier.
+
+To mitigate this, we should ensure that the full public key is compared
+to the recorded one when authenticating a BotInstance rather than merely
+comparing the fingerprint.
+
+In addition, a more modern hashing algorithm should be used to calculate the
+fingerprint. In this case, we have selected SHA256 as this is more resistant
+compared to hash functions such as MD5 or SHA1.
+
+## Alternatives
+
+### Defer the Heartbeats work and solely improve the `token` join method
+
+We could modify the way we record the generation counter for the `token` join
+method without introducing the BotInstance resource. Instead of storing a
+single counter within the Bot User labels, we could store a JSON encoded map
+of counters.
+
+One challenge would be contention over the user resource if a large number of
+Bot instances are trying to renew their certificates at the same time. Our
+Backend has limited support for transactional consistency and this increases the
+risk of two Bot instances renewing simultaneously and producing an inconsistent
+state that locks one of them out.
+
+### Introduce renewal-less and generation-less `token` join method
+
+One option is to introduce a new join method that does not produce renewable
+certificates. There would be no need for the fragile generation counter
+and the join token would be continually re-used to join as is done for the
+delegated join methods. This also circumvents the need for a one-to-one binding
+between a Bot instance and a Bot.
+
+This token would be incredibly sensitive and if stolen, there would be no
+automated mechanisms to detect this as exists today with the generation counter.
+
+It likely makes more sense to improve the existing `token` join method rather
+than introduce a variant which behaves differently and is less secure. It would
+increase the complexity of the codebase and the user experience.
+
+## Out of Scope
+
+These tasks are out of scope of this RFD but could be considered natural
+follow-on tasks.
+
+### Multi-phase Commit of Generation Counter
+
+Currently, the generation counter is fragile as it is incremented server side
+without confirmation that `tbot` has been able to use and persist the new
+credentials. If `tbot` does not receive confirmation of the renewal or is
+unable to persist the new credentials, it will be locked out on it's next
+attempt to renew.
+
+We could introduce a multi-phase commit of the generation counter. This would
+provide more robustness to the renewal process.
+
+### Locking of Individual Bot Instances
+
+Currently, it's only possible to lock out an entire Bot user. This means that
+when managing a large fleet, it would not be able to lock out a specific host
+that had been compromised. This is likely to be a major friction point for those
```

Comment:
We can leave this out of scope but we should explore it in the future.

We can already lock any heartbeating agent based on its Host ID today, so it feels like this should extend to bots without too much trouble.

Reply:
Yeah, we should add _something_ here, but it's trivial to circumvent when the join token can just be reused, @strideynet had some ideas here: https://github.com/gravitational/teleport/pull/36510/files/61e17365cac7aaec059de86d470a57a6b612eebd#r1602421886